File: services/server/src/server-protocolo/protocol.py
import safe_socket


def uint32_to_bytes(value: int) -> bytes:
    return bytes([
        (value >> 24) & 0xFF,
        (value >> 16) & 0xFF,
        (value >> 8) & 0xFF,
        value & 0xFF,
    ])


def bytes_to_uint32(data: bytes) -> int:
    return (int(data[0]) << 24) | (int(data[1]) << 16) | (int(data[2]) << 8) | int(data[3])


def receive_frame(sock):
    header = safe_socket.recv_all(sock, 4)
    if not header:
        return None
    payload_length = bytes_to_uint32(header)
    if payload_length == 0:
        return b""
    return safe_socket.recv_all(sock, payload_length)


def send_frame(sock, payload: bytes):
    header = uint32_to_bytes(len(payload))
    safe_socket.send_all(sock, header + payload)


def deserialize_bet(payload: bytes | str) -> tuple[str, str, str, str, str, str]:
    decoded = payload.decode("utf-8") if isinstance(payload, bytes) else payload
    fields = decoded.split(",")
    if len(fields) != 6:
        raise ValueError(f"Invalid bet payload: {decoded!r}")
    return tuple(fields)


def deserialize_batch(payload: bytes) -> list[tuple[str, str, str, str, str, str]]:
    if not payload:
        return []

    bets = []
    for line in payload.decode("utf-8").splitlines():
        line = line.strip()
        if line:
            bets.append(deserialize_bet(line))
    return bets


def serialize_winner(fields: tuple[str, str, str, str, str, str]) -> str:
    _, first_name, last_name, document, birthdate, number = fields
    return f"{first_name},{last_name},{document},{birthdate},{number}"

File: services/client/src/cliente-protocolo/protocol.go
package protocol

import (
	"fmt"
	"net"
	"strings"

	"github.com/7574-sistemas-distribuidos/tp-nivelador/src/safe_socket"
)

func uint32ToBytes(value uint32) []byte {
	return []byte{
		byte(value >> 24),
		byte(value >> 16),
		byte(value >> 8),
		byte(value),
	}
}

func bytesToUint32(bytes []byte) uint32 {
	return (uint32(bytes[0]) << 24) |
		(uint32(bytes[1]) << 16) |
		(uint32(bytes[2]) << 8) |
		uint32(bytes[3])
}

func SendFrame(conn net.Conn, payload []byte) error {
	header := uint32ToBytes(uint32(len(payload)))
	return safe_socket.SendAll(conn, append(header, payload...))
}

func ReceiveFrame(conn net.Conn) ([]byte, error) {
	header, err := safe_socket.RecvAll(conn, 4)
	if err != nil {
		return nil, err
	}
	payloadLength := bytesToUint32(header)
	if payloadLength == 0 {
		return nil, nil
	}
	return safe_socket.RecvAll(conn, int(payloadLength))
}

func SerializeBatch(agencyID string, rows []string) ([]byte, error) {
	if len(rows) == 0 {
		return nil, nil
	}

	var builder strings.Builder
	for i, row := range rows {
		fields := strings.Split(row, ",")
		if len(fields) != 5 {
			return nil, fmt.Errorf("invalid bet row %q", row)
		}
		if i > 0 {
			builder.WriteByte('\n')
		}
		builder.WriteString(agencyID)
		builder.WriteByte(',')
		builder.WriteString(strings.Join(fields, ","))
	}
	return []byte(builder.String()), nil
}
